simulator/mips/mips_instr_decode.h

```diff
@@ -0,0 +1,95 @@
+ /* mips_instr_decode.h - instruction decoder for mips
+ * @author Andrey Agrachev agrachev.af@phystech.edu
+ * Copyright 2018 MIPT-MIPS
+ */
+
+#ifndef MIPS_INSTR_DECODE_H
+#define MIPS_INSTR_DECODE_H
+
+enum instr_size_and_shifts {
+    FUNCT_SIZE    = 6,
+    SHAMT_SIZE    = 5,
+    RD_SIZE       = 5,
+    RT_SIZE       = 5,
+    RS_SIZE       = 5,
+    OPCODE_SIZE   = 6,
+    ASI_IMM_SIZE  = 16,
+    ASJ_IMM_SIZE  = 26,
+
+    FUNCT_SHIFT   = 0,
+    SHAMT_SHIFT   = FUNCT_SHIFT + FUNCT_SIZE,
+    RD_SHIFT      = SHAMT_SHIFT + SHAMT_SIZE,
+    RT_SHIFT      = RD_SHIFT    + RD_SIZE,
+    RS_SHIFT      = RT_SHIFT    + RT_SIZE,
+    OPCODE_SHIFT  = RS_SHIFT    + RS_SIZE,
+    ASI_IMM_SHIFT = 0,
+    ASJ_IMM_SHIFT = 0,
+};
+
+//32bits                                        01234567'01234567'01234567'01234567
+static const constexpr uint32 FUNCT_MASK    = 0b00000000'00000000'00000000'00111111;
+static const constexpr uint32 SHAMT_MASK    = 0b00000000'00000000'00000111'11000000;
+static const constexpr uint32 RD_MASK       = 0b00000000'00000000'11111000'00000000;
+static const constexpr uint32 RT_MASK       = 0b00000000'00011111'00000000'00000000;
+static const constexpr uint32 RS_MASK       = 0b00000011'11100000'00000000'00000000;
+static const constexpr uint32 OPCODE_MASK   = 0b11111100'00000000'00000000'00000000;
+static const constexpr uint32 ASI_IMM_MASK  = 0b00000000'00000000'11111111'11111111;
+static const constexpr uint32 ASJ_IMM_MASK  = 0b00000011'11111111'11111111'11111111;
+
+struct _MIPSInstrDecode
```

Comment:
Please remove the underscore. According to C++ standard, names starting with underscore and capital letter cause undefined behavior.